mlir/test/Conversion/ArithToEmitC/arith-to-emitc.mlir

```diff
@@ -390,3 +390,94 @@ func.func @arith_int_to_float_cast_ops(%arg0: i8, %arg1: i64) {
 
   return
 }
+
+// -----
+
+func.func @arith_cmpi_eq(%arg0: i32, %arg1: i32) -> i1 {
+  // CHECK-LABEL: arith_cmpi_eq
+  // CHECK-SAME: ([[Arg0:[^ ]*]]: i32, [[Arg1:[^ ]*]]: i32)
+  // CHECK-DAG: [[EQ:[^ ]*]] = emitc.cmp eq, [[Arg0]], [[Arg1]] : (i32, i32) -> i1
+  %eq = arith.cmpi eq, %arg0, %arg1 : i32
+  // CHECK: return [[EQ]]
+  return %eq: i1
+}
+
+func.func @arith_cmpi_ult(%arg0: i32, %arg1: i32) -> i1 {
+  // CHECK-LABEL: arith_cmpi_ult
+  // CHECK-SAME: ([[Arg0:[^ ]*]]: i32, [[Arg1:[^ ]*]]: i32)
+  // CHECK-DAG: [[CastArg0:[^ ]*]] = emitc.cast [[Arg0]] : i32 to ui32
+  // CHECK-DAG: [[CastArg1:[^ ]*]] = emitc.cast [[Arg1]] : i32 to ui32
+  // CHECK-DAG: [[ULT:[^ ]*]] = emitc.cmp lt, [[CastArg0]], [[CastArg1]] : (ui32, ui32) -> i1
+  %eq = arith.cmpi ult, %arg0, %arg1 : i32
+  // CHECK: return [[ULT]]
+  return %eq: i1
+}
+
+// -----
+
+func.func @trunci(%arg0: i32) -> i8 {
+  // CHECK-LABEL: trunci
+  // CHECK-SAME: (%[[Arg0:[^ ]*]]: i32)
+  // CHECK: %[[CastUI:.*]] = emitc.cast %[[Arg0]] : i32 to ui32
+  // CHECK: %[[Trunc:.*]] = emitc.cast %[[CastUI]] : ui32 to ui8
+  // CHECK: emitc.cast %[[Trunc]] : ui8 to i8
+  %truncd = arith.trunci %arg0 : i32 to i8
+
+  return %truncd : i8
+}
+
+// -----
+
+func.func @index_cast(%arg0: i32) -> i32 {
+  // CHECK-LABEL: index_cast
+  // CHECK-SAME: ([[Arg0:[^ ]*]]: i32)
+  // CHECK: %[[ICast:.*]] = emitc.cast [[Arg0]] : i32 to index
```

Comment:
The spec for index_cast says `If casting to a wider integer, the value is sign-extended. If casting to a narrower integer, the value is truncated.`. Considering that on C level, `i32` is signed and `index` (as `size_t`) is unsigned, does this cast perform a sign-extension?

Reply:
Good catch: if the C representative of `index` (`size_t`) is at least as wide as `int`, then the cast will be unsigned and not perform sign extension. Again, had we known the bitwidth of `size_t`, we wouldn't face that issue; I've avoided so far to use a test, but we may indeed have to.

Reply:
This seems to be covered by 
If the **destination type is unsigned**, the resulting value is the smallest unsigned value equal to the source value modulo 2^n where n is the number of bits used to represent the destination type. That is [...] **signed integers are sign-extended**.